src/executor.rs

```diff
@@ -0,0 +1,45 @@
+use futures::future::{ExecuteError, Executor, Future};
+use std::sync::Arc;
+
+pub trait PulsarExecutor: Executor<BoxSendFuture> + Send + Sync + 'static + Clone {
```

Comment:
So I don't think I originally added clone here because the idea was to quickly convert this trait into `TaskExecutor` then use its arc to be cloneable. So maybe it makes sense to reduce a lot of the passing the generic executor around in favor of quickly changing it to the concrete type of `TaskExecutor`.

Reply:
Yeah sorry, I think I just added that to get to the next error and forgot to fix it later. So we want all parts of that touch the public API to expose the polymorphic interface but everything else internally can use `TaskExecutor`?

Reply:
Yup, thats the idea :)

Reply:
Ah! It just clicked what the trick was, two questions,

If `Connection`, for example, holds a `TaskExecutor` 
```rust
pub struct Connection {
    addr: String,
    sender: ConnectionSender,
    sender_shutdown: Option<oneshot::Sender<()>>,
    receiver_shutdown: Option<oneshot::Sender<()>>,
    executor: TaskExecutor,
}
```

We will have a method `new<E: PulsarExecutor>(...)`, but also a method `fn executor(&self) -> TaskExecutor`, right? In that case we may be leaking the internal type. I can change it to `-> impl PulsarExecutor`?

The other thing, calling `Arc::new(executor)`, that allocates a new `Arc` every time we need to get a reference to `TaskExecutor`? If that is in `Connection::new` doesn't that mean every new connection will allocate a new `Arc`?

That was more than two, sorry!

Reply:
I think its fine to expose TaskExecutor for now, I think this crate is getting an upgrade so not a big deal.

I think allocating a new Arc per connection is fine for now but if we are getting `TaskExecutor` maybe it makes sense to create a new fn that takes a `TaskExecutor`?

Reply:
Don't we want to expose `new` with the `PulsarExecutor` trait so that `vector` is able to implement it? Let me know what you think of the PR as it is now, appreciate your time

Reply:
We would need to pass an executor that is generic, it probably should be public but there is blanket impl so nothing can implement that trait but they must implement the proper traits. So vectors executor already implements this trait, ill take a peek at this tomorrow. Thanks!

Reply:
Yes, you're right. I forgot about the blanket impl. Let me know if there are any other changes you'd like to see. I'm open to helping `pulsar` update to `std::futures` also if you folks need any help.

Reply:
Yeah, I think for now we don't need std::futures support but down the line when we update well prob want that.